### DISCUSSION TITLE
SDK php e_invoice

### DISCUSSION THREAD

**Discussion by Garagulp** [fattureincloud/api#47] [Q&A] [answered]
Per emettere una fattura elettronica con SDK php ho letto che bisogna settare true il parametro e_invoice. 
Non riesco a capire come vada fatto. Ho pensato a soluzioni di questo tipo, ma sono sbagliate di sicuro ...
Grazie

```
$entity = new Entity;
$entity
->setId(1)
->setName("$nome $cognome")
->sete_invoice(true)

// Oppure

$invoice->sete_invoice(true)
```

**Comment by valmoz** [comment]
Ciao @Garagulp 
il metodo è **setEInvoice($e_invoice)**.  Credo che l'autocomplete del tuo IDE possa darti una mano nella ricerca di queste informazioni.
Ti ricordo che puoi sempre controllare il [readme della libreria](https://github.com/fattureincloud/fattureincloud-php-sdk) per avere maggiori informazioni sui metodi disponibili.

Documentazione utile:
- [PHP SDK Github Repository](https://github.com/fattureincloud/fattureincloud-php-sdk)
- [Invoice Creation: E-Invoice](https://developers.fattureincloud.it/docs/guides/invoice-creation/#3%EF%B8%8F%E2%83%A3-step-three-e-invoice)
- [E-Invoice Management](https://developers.fattureincloud.it/docs/guides/e-invoice-management)

**Comment by Garagulp** [comment, marked as answer]
Fantastico, funziona. Grazie!! Riporto il codice così chi non è un professionista come me e si imbatte in questo thread capisce subito come fare.

```
$invoice = new IssuedDocument;
$invoice->setEInvoice(TRUE);
$invoice->setType(IssuedDocumentType::INVOICE);
```
Ti devo disturbare per un ultimo problema. La fattura elettronica viene generata correttamente. Nel vostro portale ho provato ad eseguire la verifica formale del documento e viene trovato l'errore "Nei dettagli di pagamento, il campo ModalitaPagamento non è nel formato valido".

Riporto il codice della modalità di pagamento, è l'unica modalità che ho inserito
```
$invoice->setPaymentMethod(
	new PaymentMethod(
		array(
			"id" => 1980136
		)
	)
);
```
L'id corrisponde ad una modalità che ho chiamato PayPal, non ho inserito altri parametri. Il pagamento è anticipato ed emetto la fattura dopo che l'ho incassato.

Per risolvere il problema è sufficiente che nel vostro portale clicco su Modifica, poi Finalizza senza modificare nessun parametro. Ripeto la Verifica formale e non vengono trovati errori.

Conoscete il motivo? Vorrei automatizzare in un secondo momento tutto il processo e devo prima riuscire a risolvere questo problema.
Grazie infinite.

**Reply by valmoz** [reply]
Ciao @Garagulp hai controllato la guida [E-Invoice XML Customization](https://developers.fattureincloud.it/docs/guides/e-invoice-xml-customisation)? Devi impostare ModalitaPagamento.

Vedi anche [questa discussione precedente a riguardo](https://github.com/fattureincloud/api/discussions/25)

**Reply by Garagulp** [reply]
Ciao valmoz, avevo letto anche quella guida, ma non l'avevo ben capita. Dovreste aggiungere qualche esempio, almeno per quanto riguarda la ModalitaPagamento, che a quanto vedo è fondamentale per le fatture elettroniche.
Ho scritto questo codice ma non funziona

```
$invoice->setEiData(
	new IssuedDocumentEiData(
		array(
			"paymentMethod" => "MP05"
		)
	)
);
```
Purtroppo non sono molto bravo con la programmazione a oggetti. Puoi dirmi dove sbaglio? Scusa ancora per il disturbo.

**Reply by emajo** [reply]
Ciao @Garagulp così dovrebbe funzionare: 

```
$invoice->setEiData(
    new IssuedDocumentEiData(
        array(
            "payment_method" => "MP05"
        )
    )
);
```

aggiungeremo sicuramente qualche esempio

Emanuele

**Reply by Garagulp** [reply]
> I dati del documento sono formalmente corretti.

Ciao Emanuele, grazie infinite, avete un codice fenomenale. Erano anni che non programmavo e, in vacanza al mare tra un bagno e l'altro, con il vostro fondamentale aiuto, ho completato l'integrazione del mio piccolo sito. Grazie!
